src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,218 @@
+    ///
+    /// Each source code file is stored in the source map in the compiler and has a
+    /// `lo` and a `hi` (lowest and highest bytes in this source map which can be seen as one huge
+    /// string to simplify things). So in this case, this represents the starting byte of the
+    /// current file. It'll be used later on to retrieve the "definition span" from the
+    /// `span_correspondance_map` (which is inside `context`).
+    ///
+    /// This when we transform the "span" we have from reading the input into a "span" which can be
+    /// used as index to the `span_correspondance_map` to get the definition of this item.
+    ///
+    /// So in here, `file_span_lo` is representing the "lo" byte in the global source map, and to
+    /// make our "span" works in there, we simply add `file_span_lo` to our values.
+    crate fn new_in_file(file_span_lo: u32, lo: u32, hi: u32) -> Self {
+        Self { lo: lo + file_span_lo, hi: hi + file_span_lo }
+    }
+
+    crate fn empty() -> Self {
+        Self { lo: 0, hi: 0 }
+    }
+
+    /// Extra the `lo` and `hi` from the [`Span`] and discard the unused syntax context.
+    fn new_from_span(sp: Span) -> Self {
+        Self { lo: sp.lo().0, hi: sp.hi().0 }
+    }
+}
+
+/// This function will do at most two things:
+///
+/// 1. Generate a `span` correspondance map which links an item `span` to its definition `span`.
+/// 2. Collect the source code files.
+///
+/// It returns the `krate`, the source code files and the `span` correspondance map.
+///
+/// Note about the `span` correspondance map: the keys are actually `(lo, hi)` of `span`s. We don't
+/// need the `span` context later on, only their position, so instead of keep a whole `Span`, we
+/// only keep the `lo` and `hi`.
+crate fn collect_spans_and_sources(
+    tcx: TyCtxt<'_>,
+    krate: clean::Crate,
+    src_root: &Path,
+    include_sources: bool,
+    generate_link_to_definition: bool,
+) -> (clean::Crate, FxHashMap<PathBuf, String>, FxHashMap<LightSpan, LinkFromSrc>) {
+    let mut visitor = SpanMapVisitor { tcx, matches: FxHashMap::default() };
+
+    if include_sources {
+        if generate_link_to_definition {
+            intravisit::walk_crate(&mut visitor, tcx.hir().krate());
+        }
+        let (krate, sources) = sources::collect_local_sources(tcx, src_root, krate);
+        (krate, sources, visitor.matches)
+    } else {
+        (krate, Default::default(), Default::default())
+    }
+}
+
+struct SpanMapVisitor<'tcx> {
+    crate tcx: TyCtxt<'tcx>,
+    crate matches: FxHashMap<LightSpan, LinkFromSrc>,
+}
+
+impl<'tcx> SpanMapVisitor<'tcx> {
+    /// This function is where we handle `hir::Path` elements and add them into the "span map".
+    ///
+    /// It returns `false` if we don't want to recurse in this path (in macros for example).
+    fn handle_path(&mut self, path: &rustc_hir::Path<'_>, path_span: Option<Span>) -> bool {
```

Comment:
Is the return value ever actually used? I searched in the diff for `handle_path` and couldn't find any uses of the return value.

Reply:
It isn't anymore, good catch!